lib/nixos-system.nix

```diff
@@ -1,43 +1,93 @@
 { self, lib, home-manager }:
 
-{ modules ? [ ], globalSpecialArgs ? { }, nixosSpecialArgs ? { }, hmSpecialArgs ? { }, ... } @ args:
-lib.nixosSystem (lib.recursiveUpdate (removeAttrs args [ "globalSpecialArgs" "nixosSpecialArgs" "hmSpecialArgs" ]) {
+with lib;
+
+{
+  # The global modules are included in both NixOS and home-manager.
+  globalModules ? [ ]
+
+  # Home-manager specific modules.
+, hmModules ? [ ]
+
+  # NixOS specific modules.
+, nixosModules ? [ ]
+
+  # The global extra arguments are included in both NixOS and home-manager.
+, globalSpecialArgs ? { }
+
+  # Home-manager specific extra arguments.
+, hmSpecialArgs ? { }
+
+  # NixOS specific extra arguments.
+, nixosSpecialArgs ? { }
+
+, ...
+} @ args:
+let
+  args' = removeAttrs args [
+    "globalModules"
+    "nixosModules"
+    "hmModules"
+
+    "globalSpecialArgs"
+    "nixosSpecialArgs"
+    "hmSpecialArgs"
+  ];
+in
+nixosSystem (recursiveUpdate args' {
   specialArgs = {
-    mode = "NixOS";
-    soxin = self;
-  } // globalSpecialArgs // nixosSpecialArgs;
+    # send home-manager down to the NixOS modules
+    inherit home-manager;
 
-  modules = modules ++ [
-    {
-      _module.args = { inherit home-manager; };
-    }
+    # the mode allows us to tell at what level we are within the modules.
+    mode = "NixOS";
 
-    self.nixosModules.soxin
+    # send soxin to all NixOS modules
+    soxin = self;
+  }
+  # include the global special arguments.
+  // globalSpecialArgs
+  # include the NixOS special arguments.
+  // nixosSpecialArgs;
```

Comment:
The priority of the ordering here looks good to me. I wonder if we should use `recursiveUpdate` though.

Reply:
recursiveUpdate may be really slow if it ends up merging `pkgs` with `pkgs`. That's why I thought this should suffice for now.

Reply:
Indeed, I forgot that you would probably pass down some `pkgs` at some point.